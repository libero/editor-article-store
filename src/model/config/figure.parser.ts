import {get} from 'lodash';
import {
  DOMParser as ProseMirrorDOMParser,
  DOMSerializer,
  Fragment,
  Node as ProsemirrorNode,
  Schema
} from 'prosemirror-model';
import xmldom from "xmldom";
import { getLicenseUrl } from '../figure';

const MISSING_NODES_SELECTORS_MAP = {
  figureTitle: 'fig > caption > title',
  figureLegend: 'fig > caption > p',
  figureAttribution: 'fig > attrib'
};

export function parseFigure(dom: Element, schema: Schema): Fragment {
  const figureContent = ProseMirrorDOMParser.fromSchema(schema).parseSlice(dom).content;
  const missingNodes = Object.entries(MISSING_NODES_SELECTORS_MAP).reduce((acc, nodeEntry) => {
    const [nodeName, selector] = nodeEntry;
    if (!dom.querySelector(selector)) {
      acc.push(schema.nodes[nodeName].createAndFill() as ProsemirrorNode);
    }
    return acc;
  }, [] as ProsemirrorNode[]);
  const content = figureContent.append(Fragment.fromArray(missingNodes));
  return content;
}

export function serializeFigure(node: ProsemirrorNode): Element {
  const xmlDoc = new xmldom.DOMImplementation().createDocument('', '', null);
  const figure = xmlDoc.createElement('fig');

  const label = xmlDoc.createElement('label');
  label.appendChild(xmlDoc.createTextNode(node.attrs.label));
  figure.appendChild(label);

  const graphic = xmlDoc.createElement('graphic');
  graphic.setAttribute('mime-subtype', get(node.attrs.img.match(/\.([^\.]+)$/), '1', ''));
  graphic.setAttribute(' xlink:href', node.attrs.img);
  graphic.setAttribute('mimetype', 'image');
  figure.appendChild(graphic);

  const caption = xmlDoc.createElement('caption');
  const title = findNodes(node, 'figureTitle')[0];
  caption.appendChild(serializeFigureNode(title, xmlDoc));
  const legend = findNodes(node, 'figureLegend')[0];
  caption.appendChild(serializeFigureNode(legend, xmlDoc));
  const attrib = findNodes(node, 'figureAttribution')[0];
  caption.appendChild(serializeFigureNode(attrib, xmlDoc));
  figure.appendChild(caption);

  findNodes(node, 'figureLicense').map(licenseNode => figure.appendChild(serializeFigureLicense(licenseNode, xmlDoc)));

  return figure;
}

function serializeFigureNode(node: ProsemirrorNode, document: Document): Node {
  const serializer = DOMSerializer.fromSchema(node.type.schema);
  return serializer.serializeNode(node, {document});
}

function serializeFigureLicense(node: ProsemirrorNode, xmlDoc: Document) {
  const permissions = xmlDoc.createElement('permissions');

  const ccStmt = xmlDoc.createElement('copyright-statement');
  ccStmt.appendChild(xmlDoc.createTextNode(node.attrs.licenseInfo.copyrightStatement));
  permissions.appendChild(ccStmt);

  const ccYear = xmlDoc.createElement('copyright-year');
  ccYear.appendChild(xmlDoc.createTextNode(node.attrs.licenseInfo.copyrightYear));
  permissions.appendChild(ccYear);

  const ccHolder = xmlDoc.createElement('copyright-holder');
  ccHolder.appendChild(xmlDoc.createTextNode(node.attrs.licenseInfo.copyrightHolder));
  permissions.appendChild(ccHolder);

  const licenseEl = xmlDoc.createElement('license');
  licenseEl.setAttribute(' xlink:href', getLicenseUrl(node.attrs.licenseInfo.licenseType));
  licenseEl.appendChild(serializeFigureNode(node, xmlDoc));
  permissions.appendChild(licenseEl);

  return permissions;
}

function findNodes(parentNode: ProsemirrorNode, nodeName: string): ProsemirrorNode[] {
  const result: ProsemirrorNode[] = [];

  parentNode.descendants((node: ProsemirrorNode, pos: number, parent: ProsemirrorNode) => {
    if (node.type.name === nodeName) {
      result.push(node);
    }
    return Boolean(node.childCount);
  });

  return result;
}
