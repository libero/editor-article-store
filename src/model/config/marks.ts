import { DOMOutputSpecArray, Node as ProsemirrorNode } from 'prosemirror-model';

export const marks = {
  subscript: {
    parseDOM: [{ tag: 'sub' }],
    toDOM(): DOMOutputSpecArray {
      return ['sub', 0];
    }
  },

  superscript: {
    parseDOM: [{ tag: 'sup' }],
    toDOM(): DOMOutputSpecArray {
      return ['sup', 0];
    }
  },

  italic: {
    parseDOM: [{ tag: 'italic' }],
    toDOM(): DOMOutputSpecArray {
      return ['italic', 0];
    }
  },

  strikethrough: {
    parseDOM: [{ tag: 'sc' }],
    toDOM(): DOMOutputSpecArray {
      return ['sc', 0];
    }
  },

  underline: {
    parseDOM: [{ tag: 'underline' }],
    toDOM(): DOMOutputSpecArray {
      return ['underline', 0];
    }
  },

  bold: {
    parseDOM: [{ tag: 'bold' }],
    toDOM(): DOMOutputSpecArray {
      return ['bold', 0];
    }
  },
  link: {
    attrs: {
      href: { default: undefined }
    },
    inclusive: false,
    parseDOM: [
      {
        tag: 'ext-link[ext-link-type="uri"]',
        getAttrs(dom: Element) {
          return { href: dom.getAttribute('xlink:href') };
        }
      }
    ],
    toDOM(node: ProsemirrorNode) {
      const { href } = node.attrs;
      return ['ext-link', { 'xlink:href': href, 'ext-link-type': 'uri' }, 0];
    }
  }
};
