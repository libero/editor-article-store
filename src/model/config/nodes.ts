import {DOMOutputSpec, Node as ProsemirrorNode} from 'prosemirror-model';
import { v4 as uuidv4 } from 'uuid';
import { getTextContentFromPath } from '../utils';
import {
  createEmptyLicenseAttributes,
  createFigureLicenseAttributes,
  getFigureImageUrlFromXml
} from '../figure';
import {parseFigure, serializeFigure} from './figure.parser';

function getTitleLevel(title: Element): number {
  let parent = title.parentNode;
  let level = 0;
  while (parent && parent.nodeName.toLowerCase() !== 'body') {
    if (parent.nodeName.toLowerCase() === 'sec') {
      level++;
    }
    parent = parent.parentNode;
  }
  return level;
}

export const nodes = {
  doc: {
    content: 'block+'
  },

  annotatedReferenceInfoDoc: {
    group: 'block',
    content: 'inline*',
    parseDOM: [{ tag: 'article-title' }, { tag: 'source' }],
    toDOM() {
      return ['p', 0];
    }
  },

  'article-title': {
    group: 'block',
    content: 'inline*',
    parseDOM: [{ tag: 'article-title' }],
    toDOM() {
      return ['article-title', 0];
    }
  },

  abstract: {
    group: 'block',
    content: 'paragraph',
    parseDOM: [{ tag: 'abstract' }],
    toDOM() {
      return ['p', 0];
    }
  },

  keyword: {
    content: 'text*',
    atom: true,
    parseDOM: [{ tag: 'kwd' }],
    toDOM() {
      return ['div', 0];
    }
  },

  paragraph: {
    content: 'inline*',
    group: 'block',
    parseDOM: [{ tag: 'p' }],
    toDOM() {
      return ['p', 0];
    }
  },

  section: {
    content: 'block*',
    group: 'block',
    parseDOM: [{ tag: 'sec' }],
    toDOM() {
      return ['section', 0];
    }
  },

  heading: {
    attrs: {
      level: { default: 1 },
      domId: { default: uuidv4() }
    },
    content: 'inline*',
    defining: true,
    group: 'block',
    parseDOM: [
      {
        tag: 'sec > title',
        getAttrs(dom: Element) {
          return {
            level: getTitleLevel(dom),
            domId: uuidv4()
          };
        }
      },
      { tag: 'h1', attrs: { level: 1, domId: uuidv4() } },
      { tag: 'h2', attrs: { level: 2, domId: uuidv4() } },
      { tag: 'h3', attrs: { level: 3, domId: uuidv4() } },
      { tag: 'h4', attrs: { level: 4, domId: uuidv4() } }
    ],
    toDOM(node: ProsemirrorNode) {
      return ['h' + node.attrs.level, { id: node.attrs.domId }, 0];
    }
  },

  refCitation: {
    content: undefined,
    attrs: {
      refId: { default: null },
      refText: { default: null }
    },
    group: 'inline',
    atom: true,
    inline: true,
    parseDOM: [
      {
        tag: 'xref[ref-type="bibr"]',
        getAttrs(dom: Element) {
          return {
            refText: dom.textContent,
            refId: dom.getAttribute('rid')
          };
        }
      },
      {
        tag: 'a.citation[data-cit-type="reference"]',
        priority: 100,
        getAttrs(dom: Element) {
          return {
            refText: dom.getAttribute('data-ref-text'),
            refId: dom.getAttribute('data-ref-id')
          };
        }
      }
    ],
    toDOM(node: ProsemirrorNode): DOMOutputSpec {
      return [
        'a',
        {
          href: '#',
          class: 'citation',
          'data-cit-type': 'reference',
          'data-ref-id': node.attrs.refId,
          'data-ref-text': node.attrs.refText
        }
      ];
    }
  },

  boxText: {
    content: 'paragraph*',
    group: 'block',
    parseDOM: [{ tag: 'boxed-text' }],
    toDOM() {
      return ['boxed-text', 0];
    }
  },

  figure: {
    content: '(figureTitle | figureLegend | figureAttribution | figureLicense)*',
    group: 'block',
    atom: true,
    draggable: true,
    attrs: {
      id: { default: '' },
      label: { default: '' },
      img: { default: '' }
    },
    parseDOM: [
      {
        tag: 'fig',
        getAttrs(dom: Element) {
          return {
            id: dom.getAttribute('id'),
            label: getTextContentFromPath(dom, 'label') || '',
            img: getFigureImageUrlFromXml(dom)
          };
        },
        getContent: parseFigure
      }
    ],
    toDOM: serializeFigure
  },

  figureTitle: {
    content: 'inline*',
    parseDOM: [{ tag: 'caption > title' }],
    toDOM() {
      return ['title', 0];
    }
  },

  figureLegend: {
    content: 'inline*',
    parseDOM: [
      { tag: 'caption > p', priority: 100 },
      { tag: 'label', ignore: true }
    ],
    context: 'figure',
    toDOM() {
      return ['p', 0];
    }
  },

  figureAttribution: {
    content: 'inline*',
    parseDOM: [
      { tag: 'attrib', priority: 100 },
      { tag: 'label', ignore: true }
    ],
    context: 'figure',
    toDOM() {
      return ['p', 0];
    }
  },

  figureLicense: {
    content: 'inline*',
    group: 'block',
    attrs: {
      licenseInfo: { default: createEmptyLicenseAttributes() }
    },
    contentElement: 'license > license-p',
    parseDOM: [
      {
        tag: 'permissions',
        getAttrs(dom: Element) {
          return { licenseInfo: createFigureLicenseAttributes(dom) };
        }
      },
      { tag: 'label', ignore: true }
    ],
    toDOM() {
      return ['license-p', 0];
    }
  },

  orderedList: {
    group: 'block',
    content: 'listItem+',
    parseDOM: [{ tag: 'list[list-type=order]' }],
    toDOM() {
      return ['ol', 0];
    }
  },

  bulletList: {
    group: 'block',
    content: 'listItem+',
    parseDOM: [{ tag: 'list[list-type=bullet]' }],
    toDOM() {
      return ['ul', 0];
    }
  },

  listItem: {
    content: 'paragraph block*',
    parseDOM: [{ tag: 'list-item' }],
    toDOM() {
      return ['li', 0];
    },
    defining: true
  },

  figureCitation: {
    content: 'text*',
    group: 'inline',
    attrs: {
      figIds: { default: null }
    },
    inline: true,
    parseDOM: [
      {
        tag: 'xref[ref-type="fig"]',
        getAttrs(dom: Element) {
          return {
            figIds: dom.getAttribute('rid')!.split(' ').filter(Boolean)
          };
        }
      },
      {
        tag: 'a.citation[data-cit-type="figure"]',
        priority: 100,
        getAttrs(dom: Element) {
          return {
            figIds: dom.getAttribute('data-fig-ids')!.split(' ')
          };
        }
      }
    ],
    toDOM(node: ProsemirrorNode): DOMOutputSpec {
      return [
        'a',
        {
          href: '#',
          class: 'citation',
          'data-cit-type': 'figure',
          'data-fig-ids': node.attrs.figIds.join(' ')
        },
        0
      ];
    }
  },

  text: {
    group: 'inline'
  }
};

export type NodesType = keyof typeof nodes;
