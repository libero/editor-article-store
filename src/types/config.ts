export type Config = {
  port?: number;
  awsRegion?: string;
  awsAccessKey?: string;
  awsSecretAccessKey?: string;
  awsBucketInputEventQueueUrl?: string;
  awsEndpoint?: string;
  dbEndpoint?: string;
  dbUriQuery?: string;
  dbUser?: string;
  dbPassword?: string;
  dbSSLValidate?: boolean;
  dbName?: string;
  editorS3Bucket?: string;
  srcS3Bucket?: string;
};
