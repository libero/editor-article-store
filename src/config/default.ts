import { Config } from "../types/config";

// An initial default config
export const defaultConfig: Config = {
  port: 8080,
  awsRegion: "us-east-1",
  awsAccessKey: "test",
  awsSecretAccessKey: "test",
  awsBucketInputEventQueueUrl: "http://localhost:4566/000000000000/KryiaUploadQueue",
  awsEndpoint: "http://localhost:4566",
  dbEndpoint: "mongo:27017",
  dbUriQuery: '',
  dbUser: 'root',
  dbPassword: 'password',
  dbSSLValidate: false,
  dbName: "editor",
  editorS3Bucket: "editor",
  srcS3Bucket: "kryia",
  importTransformUrl: 'http://transformer:8984/v1tov2',
  importTransformEnabled: false
};
