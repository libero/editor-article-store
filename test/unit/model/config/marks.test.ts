import { Node as ProsemirrorNode } from 'prosemirror-model';

import { marks } from '../../../../src/model/config/marks';
import * as xmldom from '@xmldom/xmldom';

describe('nodes spec', () => {
    it('checks nodes spec definition', () => {
        expect(marks).toMatchSnapshot();
    });

    it('checks sub spec', () => {
        expect(marks['subscript'].toDOM()).toEqual(['sub', 0]);
    });

    it('checks sup spec', () => {
        expect(marks['superscript'].toDOM()).toEqual(['sup', 0]);
    });

    it('checks italic spec', () => {
        expect(marks['italic'].toDOM()).toEqual(['italic', 0]);
    });

    it('checks strikethrough spec', () => {
        expect(marks['strikethrough'].toDOM()).toEqual(['sc', 0]);
    });

    it('checks underline spec', () => {
        expect(marks['underline'].toDOM()).toEqual(['underline', 0]);
    });

    it('checks bold spec', () => {
        expect(marks['bold'].toDOM()).toEqual(['bold', 0]);
    });

    it('checks link spec attributes', () => {
        const xmlDoc = new xmldom.DOMImplementation().createDocument(null, null);
        const node = xmlDoc.createElement('a');
        node.setAttribute('xlink:href', 'URL');
        expect(marks['link'].parseDOM[0].getAttrs!(node)).toEqual({ href: 'URL' });
    });

    it('checks link spec rendering', () => {
        const node = new ProsemirrorNode();
        node.attrs = { href: 'URL' };
        expect(marks['link'].toDOM(node)).toEqual(['ext-link', { 'xlink:href': 'URL', 'ext-link-type': 'uri' }, 0]);
    });
});
