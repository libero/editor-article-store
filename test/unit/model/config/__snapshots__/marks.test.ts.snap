// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodes spec checks nodes spec definition 1`] = `
Object {
  "bold": Object {
    "parseDOM": Array [
      Object {
        "tag": "bold",
      },
    ],
    "toDOM": [Function],
  },
  "italic": Object {
    "parseDOM": Array [
      Object {
        "tag": "italic",
      },
    ],
    "toDOM": [Function],
  },
  "link": Object {
    "attrs": Object {
      "href": Object {
        "default": undefined,
      },
    },
    "inclusive": false,
    "parseDOM": Array [
      Object {
        "getAttrs": [Function],
        "tag": "ext-link[ext-link-type=\\"uri\\"]",
      },
    ],
    "toDOM": [Function],
  },
  "strikethrough": Object {
    "parseDOM": Array [
      Object {
        "tag": "sc",
      },
    ],
    "toDOM": [Function],
  },
  "subscript": Object {
    "parseDOM": Array [
      Object {
        "tag": "sub",
      },
    ],
    "toDOM": [Function],
  },
  "superscript": Object {
    "parseDOM": Array [
      Object {
        "tag": "sup",
      },
    ],
    "toDOM": [Function],
  },
  "underline": Object {
    "parseDOM": Array [
      Object {
        "tag": "underline",
      },
    ],
    "toDOM": [Function],
  },
}
`;
